Feat/accountview

crypblizz8 opened the pull request:
![Demo](https://user-images.githubusercontent.com/45455218/190993945-469bf567-2392-4414-a535-394631bfa25a.gif)

Overview
- Implement the UI Button for `AccountButton` and `AccountView`
- AccountButton displays: `balance`, `ensAvatar` or `Zorb`, `address`
- AccountView displays: same features as above + `disconnect` + `copyClipboard` functions

Considerations
- In `getENSAvatar()` and `getBalance()`, should look to add them to AccountCtrl so it is only fetched once. However this may apply to other functions throughout the codeBase so needs discussion.
- Consider loading states / completed state for `copyClipboard` + ens/zorb Avatars

Other
- Long commit list as I had to rebase + move my other PR here.

IljaDaderko commented on packages/ui/src/components/w3m-modal-router/index.ts:
```diff
@@ -34,8 +35,8 @@ export class W3mModalRouter extends LitElement {
   public firstUpdated() {
-    this.resizeObserver = new ResizeObserver(([conetnt]) => {
-      const newHeight = `${conetnt.borderBoxSize[0].blockSize}px`
+    this.resizeObserver = new ResizeObserver(([content]) => {
```
Nice catch!

IljaDaderko commented on packages/ui/src/components/w3m-zorb-ens-image/styles.ts:
```diff
@@ -0,0 +1,19 @@
+    width: 60px;
+    height: 60px;
+    margin-bottom: 10px;
+  }
```
Could you add line break between css blocks for better readability please

IljaDaderko commented on packages/ui/src/components/w3m-zorb-ens-image/index.ts:
```diff
@@ -0,0 +1,62 @@
+  // -- private ----------------------------------------------------- //
+  private async getENSAvatar() {
+    try {
+      const opts: FetchEnsAvatarOpts = {
```
You shouldn't really need to specify `: FetchEnsAvatarOpts` here, as we try to keep everything to typescripts auto-inheritance i.e. if it is wrong it should be highlighted on `.fetchEnsAvatar(opts)` level

IljaDaderko commented on packages/ui/src/partials/w3m-account-button/index.ts:
```diff
@@ -1,16 +1,100 @@
+  // -- lifecycle ---------------------------------------------------- //
+  public constructor() {
+    super()
+    this.getAccounts()
```
This one is gonna be tricky, but I'll try to outline proper setps best I can:

1. Ideally we should have all of this data (acc, ens and balance) available here before the component is mounted. This can be done inside `chains/ethereum/src/utilities` `onClientConnected` function. And should also happen in `Promise.all`

2. Doing it in the above way also means that we probably need to move all `ConnectModalCtrl.closeModal` calls to `onClientConnected` as well, so that it closes after all the data is ready

3. Inside w3m-account-button component you would then need to not only access the data, but also subscribe to it (account can change if user changes it on their mobile wallet for example). You can see example of subscribing and unsubscribing to state in component like `w3m-modal-router`

IljaDaderko commented on packages/ui/src/partials/w3m-account-button/styles.ts:
```diff
@@ -60,12 +118,28 @@ export function dynamicStyles() {
-    svg path {
-      fill: ${foreground.inverse};
-    }
+    // svg path {
```
Will this be un-commented latter or can it be removed? If it is used latter on, could you add TODO here please

IljaDaderko commented on packages/ui/src/utils/Svgs.ts:
```diff
@@ -451,3 +451,77 @@ export const DEFI_IMG = svg`
   </svg>
 `
+
+export const ETH_LOGO = svg`
```
For all new svg's could you please run them through https://jakearchibald.github.io/svgomg/ and cleanup unecessary data i.e. `fill="none" xmlns="http://www.w3.org/2000/svg"` as these are inlined into final bundle.

crypblizz8 replied:
Fixed.

IljaDaderko commented on packages/ui/src/utils/UiHelpers.ts:
```diff
@@ -63,3 +63,27 @@ export async function preloadImage(src: string) {
+  }
+}
+
+export const ZorbPackageScript =
```
Any 3rd party lib we use, should come from npm. If it is not available there, you can copy source to utils i.e `Zorb.ts` (see how this was done for Whatmesh). And then import it where it is required (see if you can strip it down as well for unused code). This means that `useScript` util will also not be required.

crypblizz8 replied:
Removed + rebuilt with one small `tinycolors` library !

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,135 @@
+  // -- lifecycle ---------------------------------------------------- //
+  public constructor() {
+    super()
+    this.getAccounts()
```
Same as above. This data should be available beforehand and subscribed to.

IljaDaderko reviewed (commented):
Good stuff! I left few comments, I think biggest change that's needed is related to those data flows I mentioned and subscribing to them.

Once that's changed I'll test ui out 👍

crypblizz8 commented on package.json:
```diff
@@ -25,18 +25,19 @@
+    "@typescript-eslint/parser": "5.38.0",
     "eslint": "8.23.1",
     "eslint-config-prettier": "8.5.0",
+    "lerna": "5.5.2",
```
Think it was just my prettier re-order. Happy to revert

crypblizz8 commented:
### Overview of whole PR
- Create `AccountButton`
- Create `AccountView`
- Create `zorb-ens component`
- Altered data flow to be higher order in the lifecycle
- Added new SVGs
- Added ```tinycolor2``` package

### Recent Changes from last PR check
DataFlow
- This was done by moving up onChange events higher in the order in ```wagmiHelpers``` with ```getBalance``` and ```getENSAvatar``` with a Promise.all
- This will provide initial state data
- In scenario of the change of accounts (I.e. Account 1 --> Account 2 on Metamask), onConnectorChange and sets the relevant state
- New data appears on `AccountButton` and `AccountView` (regardless if model is open)

SVG
- Reworked `BACK_ICON` and old `CHEVRON` into just `CHEVRON`
- Kept `CLIPBOARD_BLUE` as it is a necessary SVG with the blue circle. Can optimize later

Zora
- Installed on library (```npm i tinycolor2```) for Zorb Colors
- Disabled TS-lint on the document
- Used the rest from the Zora `web-component-package`

Learnings
- Make this PR smaller next time.
- Consider the higher order data flow from beginning.

IljaDaderko commented on packages/ui/src/utils/Zorb.ts:
```diff
@@ -0,0 +1,209 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
```
Really don't like how much bloat this adds to modal for what it acheives. Let's drop Zorb and all of it's dependencies and instead simplify this in the following way:

There are plenty of ideas of how to generate random color from a given string outlined here:
https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript

Let's grab user's address and split it in 2 parts. Use those parts to create 2 random colors and just style div with them so it looks similar to a zorb. i.e. highlight gradient, body gradient and one of the colors used as border.

IljaDaderko commented on chains/ethereum/src/utils/wagmiHelpers.ts:
```diff
@@ -18,8 +18,49 @@ export function getClient() {
-function onConnectorChange(event: ConnectorData) {
-  if (event.account) AccountCtrl.setAddress(event.account)
+async function getBalance(account: string) {
+  const opts: GetBalanceOpts = {
```
```suggestion
  const opts = {
```

Let typescript do the work

IljaDaderko commented on chains/ethereum/src/utils/wagmiHelpers.ts:
```diff
@@ -18,8 +18,49 @@ export function getClient() {
+
+async function getENSAvatar(account: string) {
+  try {
+    const opts: FetchEnsAvatarOpts = {
```
```suggestion
    const opts = {
```

IljaDaderko commented on chains/ethereum/src/utils/wagmiHelpers.ts:
```diff
@@ -18,8 +18,49 @@ export function getClient() {
+
+  const balance = await ClientCtrl.ethereum().fetchBalance(opts)
+  // eslint-disable-next-line line-comment-position, no-inline-comments, no-warning-comments
+  // Todo: Check / Might be setting state twice..?
```
This needs to be figured out before merging and comments removed

IljaDaderko commented on chains/ethereum/src/utils/wagmiHelpers.ts:
```diff
@@ -18,8 +18,49 @@ export function getClient() {
+
+    return ''
+  } catch (error) {
+    // eslint-disable-next-line no-console
```
Show error toast instead of log

IljaDaderko commented on chains/ethereum/src/utils/wagmiHelpers.ts:
```diff
@@ -18,8 +18,49 @@ export function getClient() {
+async function onConnectorChange(event: ConnectorData) {
+  if (event.account) {
+    AccountCtrl.setAddress(event.account)
+    await getBalance(event.account)
```
Promise.all these

IljaDaderko commented on package.json:
```diff
@@ -25,18 +25,19 @@
     "rollup-plugin-minify-html-literals": "1.2.6",
-    "lerna": "5.5.2"
+    "typescript": "4.8.3",
+    "tinycolor2": "^1.4.2"
```
`tinycolor2` needs to be removed once comment about zorb is addressed

IljaDaderko commented on packages/ui/package.json:
```diff
@@ -24,6 +24,7 @@
   },
   "devDependencies": {
     "@types/qrcode": "1.5.0",
+    "@types/tinycolor2": "^1.4.3",
```
Needs to be removed once comment about Zorb is addressed

IljaDaderko commented on packages/ui/src/components/w3m-zorb-ens-image/index.ts:
```diff
@@ -0,0 +1,47 @@
+import { AccountCtrl } from '@web3modal/core'
```
File can be simplified once Zorb comment is addressed

IljaDaderko commented on packages/ui/src/partials/w3m-account-button/index.ts:
```diff
@@ -1,16 +1,78 @@
+      this.balance = AccountCtrl.state.balance
+    })
+
+    return () => this.unsubscribe
```
Why is this return needed here? You set `this.unsubscribe` correctly, but never unsubscribe actually. need to add

```ts
public disconnectedCallback() {
    this.unsubscribe?.()
  }
```

IljaDaderko commented on packages/ui/src/partials/w3m-account-button/styles.ts:
```diff
@@ -60,12 +116,32 @@ export function dynamicStyles() {
-      fill: ${foreground.inverse};
-    }
+    // svg path {
+    //   fill: ${foreground.inverse};
```
Commented code

IljaDaderko commented on packages/ui/src/utils/Svgs.ts:
```diff
@@ -452,6 +452,76 @@ export const DEFI_IMG = svg`
+  </svg>
+`
+
+export const CLIPBOARD_BLUE = svg`
```
All icons should be white. Their color is set in css via path fill.

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,116 @@
+      this.balance = AccountCtrl.state.balance
+    })
+
+    return () => this.unsubscribe
```
Same as above, this return is not needed and proper unsubscribing needs to be done in appropriate lit lifecycle method.

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,116 @@
+    return () => this.unsubscribe
+  }
+
+  private async copyClipboard() {
```
Show successful copy toast message here

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  // -- render ------------------------------------------------------- //
+  // eslint-disable-next-line no-warning-comments
```
Needs to be addressed before merge

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,116 @@
+          </div>
+        </div>
+
+        <div class="w3m-account-divider"></div>
```
Use css border instead of extra element

IljaDaderko commented on packages/ui/src/views/w3m-account-view/index.ts:
```diff
@@ -0,0 +1,116 @@
+    return html`
+      ${dynamicStyles()}
+
+      <div>
```
Is this outer div actually needed?

IljaDaderko commented:
All items marked red need to become components i.e

- [ ] `w3m-address` needs to to use existing text component, and utilise format address function
- [ ] `w3m-balance` needs to use existing text component and expose prop for token icon
- [ ] `w3m-round-button` needs to allow for icon and onClick action. Also add hover effect to it as with other buttons

<img width="414" alt="Screenshot 2022-09-23 at 09 41 12" src="https://user-images.githubusercontent.com/3154053/191905230-f39b2054-e2c0-4b8d-a06c-b9f88831613c.png">

devceline commented on packages/ui/src/utils/UiHelpers.ts:
```diff
@@ -66,6 +66,14 @@ export async function preloadImage(src: string) {
+  return parseFloat(address).toFixed(2)
+}
+
+export function formatAddress(address: string) {
+  return `${address.substring(0, 5)}...${address.slice(-5)}`
+}
```
I feel like this should have a flexible range instead of the fixed `5` but that's a nitpick

tunnckoCore replied:
It should be first 4 w/o the `0x` and the last 4.

crypblizz8 replied:
Great suggestion, will set with first 4 and last 4 as default + add in an optional range. @devceline @tunnckoCore

crypblizz8 reviewed (commented):
Have addressed the following in the latest changes

Major
- Zorb rewrite. Tried a few methods here. 1) The ENS Version 2) Gradient div one that was suggested. 3) Own adaptation. The ENS one was too large. The Gradient one didn't end up as pretty and using HSL is a much better outcome. Created own adaptation of the Zorbs + ENS + learnings from Gradient Div attempt. 
- Check if double setting state?
- Fix the Unsubscribe

- `w3m-address `component
- `w3m-balance` component
- `w3m-round-button` component

Minor
- Typescript for getOpts 
- Use css border instead of extra element 
- Toast message ERROR 
- Toast message for Clipboard.
- Is this outer div actually needed? 
- Address formatting
- Test Mobile View + Mobile Edge Cases
